File: src/BlazorUI/Bit.BlazorUI/Components/Chart/Interop/TypeScript/ChartJsInterop.ts
/// <reference path="types/Chartjs.d.ts" />   

interface ChartConfiguration extends Chart.ChartConfiguration {
    canvasId: string;
}

interface DotNetObjectReference {
    invokeMethod(methodName: string, ...args: any[]): any;
    invokeMethodAsync(methodName: string, ...args: any[]): Promise<any>;
}

// maybe the interop could be improved somehow by using generic similar to how it's done in C#
// but since we don't need typesafety in JavaScript and can't rely on it anyway (just like we
// have to invoke the C# delegate dynamically) we're not using generics.
interface IMethodHandler {
    methodName: string;
}

interface DelegateHandler extends IMethodHandler {
    handlerReference: DotNetObjectReference;
    returnsValue: boolean;
    ignoredIndices: number[];
}

class ChartJsInterop {
    BlazorCharts = new Map<string, Chart>();

    public setupChart(config: ChartConfiguration): boolean {
        if (!this.BlazorCharts.has(config.canvasId)) {
            this.wireUpCallbacks(config);

            let chart = new Chart(config.canvasId, config)
            this.BlazorCharts.set(config.canvasId, chart);

            return true;
        } else {
            return this.updateChart(config);
        }
    }

    public updateChart(config: ChartConfiguration): boolean {
        if (!this.BlazorCharts.has(config.canvasId))
            throw `Could not find a chart with the given id. ${config.canvasId}`;

        let myChart = this.BlazorCharts.get(config.canvasId);

        if (!myChart) return false;

        // Update datasets. This breaks the data-array-references; more in the function.
        this.mergeDatasets(myChart.config.data!.datasets!, config.data!.datasets!);
        // Update labels while keeping array references intact.
        this.mergeLabels(myChart.config.data!, config.data!);
        // Currently we only merge the datasets and the labels of the data subconfig but that
        // could be expanded in a similar fashion as the dataset's data (if there's a use-case).

        this.wireUpCallbacks(config);

        // This will add new options and update existing ones. Nothing is deleted.
        // Calling extend instead of merge avoids the unnecessary deep copy as
        // config.options is a brand new object (deserialized by blazor).
        Chart.helpers.extend(myChart.config.options, config.options);

        myChart.update();
        return true;
    }

    private mergeDatasets(oldDatasets: Array<Chart.ChartDataSets>, newDatasets: Array<Chart.ChartDataSets>) {
        // iterate backwards so we can remove datasets as we go
        for (let i = oldDatasets.length - 1; i >= 0; i--) {
            let sameDatasetInNewConfig = newDatasets.find(newD => newD.id === oldDatasets[i].id);
            if (sameDatasetInNewConfig === undefined) {
                // Remove dataset if it's not in the new config
                oldDatasets.splice(i, 1);
            } else {
                // This comment below would be the 'correct' way of updating the data while retaining the same reference.
                // However, there's quite a big issue with this. Chart.js actually listenes for modifications on
                // the data array and will decide on the update-animation by looking at the latest modifications.
                // Since this would clear the whole array and then add all the new data, Chart.js thinks every data
                // point is fresh and plays the same animation it plays when initially creating the chart.
                // To prevent Chart.js from doing that, we replace the reference which doesn't record any modifications.

                //oldDatasets[i].data.length = 0; // Remove old data
                //for (let j = 0; j < sameDatasetInNewConfig.data.length; j++) {
                //    // Add current data. Of course it won't be a number _and_ a ChartPoint but I don't how else to make ts happy
                //    oldDatasets[i].data.push(<number & Chart.ChartPoint>sameDatasetInNewConfig.data[j]);
                //}
                //delete sameDatasetInNewConfig.data; // Remove the array from the new dataset so it doesn't get copied in the next line

                // Merge everything, including the data-array reference.
                // As with the labels, deep copying(with helper.merge) is simply a waste here.
                Chart.helpers.extend(oldDatasets[i], sameDatasetInNewConfig);
            }
        }

        let currentIds = oldDatasets.map(dataset => dataset.id);
        newDatasets.filter(newDataset => !currentIds.includes(newDataset.id))
            .forEach(newDataset => oldDatasets.push(newDataset));

        // Currently the order isn't respected so simply reordering the datasets and calling update
        // won't do anything. You'd have to remove and readd them. Maybe this could be implemented later.
    }

    private mergeLabels(oldChartData: Chart.ChartData, newChartData: Chart.ChartData): void {
        const innerFunc = (oldLabels: Array<string | string[] | number | number[] | Date | Date[]>,
            newLabels: Array<string | string[] | number | number[] | Date | Date[]>) => {
            if (newLabels == null || newLabels.length === 0) {
                if (oldLabels) {
                    oldLabels.length = 0;
                }

                return oldLabels;
            }

            if (oldLabels == null) {
                return newLabels;
            }

            // clear existing labels
            oldLabels.length = 0;

            // add all the new labels
            for (var i = 0; i < newLabels.length; i++) {
                oldLabels.push(newLabels[i]);
            }

            return oldLabels;
        }

        // If it was null/undefined before it can't be done in-place so assignment is required.
        oldChartData.labels = innerFunc(oldChartData.labels!, newChartData.labels!);
        oldChartData.xLabels = innerFunc(oldChartData.xLabels!, newChartData.xLabels!);
        oldChartData.yLabels = innerFunc(oldChartData.yLabels!, newChartData.yLabels!);
    }

    private wireUpCallbacks(config: ChartConfiguration) {
        // Replace IMethodHandler objects with actual function (if present)
        // This should be "automated" in some way. We shouldn't have to add
        // (much) new code for a new callback.
        this.wireUpOptionsOnClick(config);
        this.wireUpOptionsOnHover(config);
        this.wireUpLegendOnClick(config);
        this.wireUpLegendOnHover(config);
        this.wireUpLegendItemFilter(config);
        this.wireUpGenerateLabels(config);
        this.wireUpTickCallback(config);
    }

    private wireUpOptionsOnClick(config: ChartConfiguration) {
        let getDefaultFunc = (type: any) => {
            let defaults = Chart.defaults[type] || Chart.defaults.global;
            return defaults?.onClick || Chart.defaults.global.onClick;
        };

        if (!config.options)
            return;

        config.options.onClick = this.getMethodHandler(<IMethodHandler>config.options.onClick, getDefaultFunc(config.type));
    }

    private wireUpOptionsOnHover(config: ChartConfiguration) {
        let getDefaultFunc = (type: any) => {
            let defaults = Chart.defaults[type] || Chart.defaults.global;
            return defaults?.onHover || Chart.defaults.global.onHover;
        };

        if (!config.options)
            return;

        config.options.onHover = this.getMethodHandler(<IMethodHandler>config.options.onHover, getDefaultFunc(config.type));
    }

    private wireUpLegendOnClick(config: ChartConfiguration) {
        let getDefaultHandler = (type: any) => {
            let chartDefaults = Chart.defaults[type] || Chart.defaults.global;
            return chartDefaults?.legend?.onClick || Chart.defaults.global.legend?.onClick;
        };

        if (!config.options?.legend)
            return;

        config.options.legend.onClick = this.getMethodHandler(<IMethodHandler>config.options.legend.onClick, getDefaultHandler(config.type));
    }

    private wireUpLegendOnHover(config: ChartConfiguration) {
        let getDefaultFunc = (type: any) => {
            let chartDefaults = Chart.defaults[type] || Chart.defaults.global;
            return chartDefaults?.legend?.onHover || Chart.defaults.global.legend!.onHover;
        };

        if (!config.options?.legend)
            return;

        config.options.legend.onHover = this.getMethodHandler(<IMethodHandler>config.options.legend.onHover, getDefaultFunc(config.type));
    }

    private wireUpLegendItemFilter(config: ChartConfiguration) {
        let getDefaultFunc = (type: any) => {
            let chartDefaults = Chart.defaults[type] || Chart.defaults.global;
            return chartDefaults?.legend?.labels?.filter || Chart.defaults.global.legend!.labels!.filter;
        };

        if (!config.options?.legend?.labels)
            return;

        config.options.legend.labels.filter = this.getMethodHandler(<IMethodHandler>config.options.legend.labels.filter, getDefaultFunc(config.type));
    }

    private wireUpGenerateLabels(config: ChartConfiguration) {
        let getDefaultFunc = (type: any) => {
            let chartDefaults = Chart.defaults[type] || Chart.defaults.global;
            return chartDefaults?.legend?.labels?.generateLabels || Chart.defaults.global.legend!.labels!.generateLabels;
        };

        if (!config.options?.legend?.labels)
            return;

        config.options.legend.labels.generateLabels = this.getMethodHandler(<IMethodHandler>config.options.legend.labels.generateLabels, getDefaultFunc(config.type));
    }

    private wireUpTickCallback(config: ChartConfiguration) {
        /* Defaults table (found out by checking Chart.defaults in console) -> everything undefined
         * Bar (scales): undefined
         * Bubble (scales): undefined
         * Pie & Doughnut: don't even have scale(s) field
         * HorizontalBar (scales): undefined
         * Line (scales): undefined
         * PolarArea (scale): undefined
         * Radar (scale): undefined
         * Scatter (scales): undefined
         */

        const assignCallbacks = (axes: any) => {
            if (axes) {
                for (var i = 0; i < axes.length; i++) {
                    if (!axes[i].ticks) continue;
                    axes[i].ticks.callback = this.getMethodHandler(axes[i].ticks.callback, undefined);
                    if (!axes[i].ticks.callback) {
                        delete axes[i].ticks.callback; // undefined != deleted, Chart.js throws an error if it's undefined so we have to delete it
                    }
                }
            }
        }

        if (config.options?.scales) {
            assignCallbacks(config.options.scales.xAxes);
            assignCallbacks(config.options.scales.yAxes);
        }

        if (config.options?.scale?.ticks) {
            config.options.scale.ticks.callback = this.getMethodHandler(<IMethodHandler>config.options.scale.ticks.callback, undefined);

            if (!config.options.scale.ticks.callback) {
                delete config.options.scale.ticks.callback; // undefined != deleted, Chart.js throws an error if it's undefined so we have to delete it
            }
        }
    }

    /**
     * Given an IMethodHandler (see C# code), it tries to resolve the referenced method.
     * It currently supports JavaScript functions, which are expected to be attached to the window object, and .Net delegates which can be
     * bound to .Net static functions, .Net object instance methods and more.
     *
     * When failing to recover a method from the IMethodHandler, it returns the default handler.
     *
     * @param handler the serialized IMethodHandler (see C# code)
     * @param defaultFunc the fallback value to use in case the method can't be resolved
     */
    private getMethodHandler(handler: IMethodHandler, defaultFunc?: Function) {
        if (handler == null) {
            return defaultFunc;
        }

        if (this.isDelegateHandler(handler)) {
            // stringify args and ignore all circular references. This means that objects of type DotNetObject will not be
            // deserialized correctly (since it's already a string when it reaches JSON.stringify in the blazor interop layer)
            // but the values passed to chart callbacks should never contain such objects anyway.
            // Also if we don't care about the value, don't bother to stringify.
            const stringifyArgs = (args: any[]) => {
                for (var i = 0; i < args.length; i++) {
                    if (handler.ignoredIndices.includes(i)) {
                        args[i] = '';
                    } else {
                        args[i] = this.stringifyObjectIgnoreCircular(args[i]);
                    }
                }

                return args;
            }

            if (!handler.returnsValue) {
                // https://stackoverflow.com/questions/59543973/use-async-function-when-consumer-doesnt-expect-a-promise
                return (...args: any[]) => handler.handlerReference.invokeMethodAsync(handler.methodName, stringifyArgs(args));
            } else {
                if (window.hasOwnProperty('MONO')) {
                    return (...args: any[]) => handler.handlerReference.invokeMethod(handler.methodName, stringifyArgs(args)); // only works on client side
                } else {
                    console.warn('Using C# delegates that return values in chart.js callbacks is not supported on ' +
                        "server side blazor because the server side dispatcher doesn't support synchronous interop calls. Falling back to default value.");

                    return defaultFunc;
                }
            }
        } else {
            if (handler.methodName == null) {
                return defaultFunc;
            }

            const namespaceAndFunc: string[] = handler.methodName.split('.');
            if (namespaceAndFunc.length !== 2) {
                return defaultFunc;
            }

            const namespace = (window as any)[namespaceAndFunc[0]];
            if (namespace == null) {
                return defaultFunc;
            }

            const func = namespace[namespaceAndFunc[1]];
            if (typeof func === 'function') {
                return func;
            } else {
                return defaultFunc;
            }
        }
    }

    private isDelegateHandler(handler: IMethodHandler): handler is DelegateHandler {
        return 'handlerReference' in handler;
    }

    private stringifyObjectIgnoreCircular(object: any) {
        const seen = new WeakSet();
        const replacer = (_name: any, value: any) => {
            if (
                typeof value === 'object' &&
                value !== null &&
                !(value instanceof Boolean) &&
                !(value instanceof Date) &&
                !(value instanceof Number) &&
                !(value instanceof RegExp) &&
                !(value instanceof String)
            ) {
                if (seen.has(value))
                    return undefined;

                seen.add(value);
            }

            return value;
        }

        return JSON.stringify(object, replacer);
    }
}

(window as any)[ChartJsInterop.name] = new ChartJsInterop();

File: src/BlazorUI/Bit.BlazorUI/Components/Chart/Interop/ChartJsInterop.cs
using System.Dynamic;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using Newtonsoft.Json.Serialization;

namespace Bit.BlazorUI;

/// <summary>
/// Interop layer from C# to JavaScript.
/// </summary>
internal static class ChartJsInterop
{
    private const string ChartJsInteropName = "ChartJsInterop";

    internal static JsonSerializerSettings JsonSerializerSettings { get; } = new JsonSerializerSettings
    {
        NullValueHandling = NullValueHandling.Ignore,
        ContractResolver = new IgnoreDatasetCountContractResolver
        {
            NamingStrategy = new CamelCaseNamingStrategy(true, false)
        },
        Converters =
        {
            new IsoDateTimeConverter()
        }
    };

    public static async Task InitChartJs(this IJSRuntime jsRuntime)
    {
        var result = await jsRuntime.InvokeAsync<string>($"{ChartJsInteropName}.initChartJs");
    }

    /// <summary>
    /// Set up a new chart. Call only once.
    /// </summary>
    /// <param name="jsRuntime"></param>
    /// <param name="chartConfig">The config for the new chart.</param>
    /// <returns></returns>
    public static ValueTask<bool> SetupChart(this IJSRuntime jsRuntime, ConfigBase chartConfig)
    {
        dynamic dynParam = StripNulls(chartConfig);
        Dictionary<string, object> param = ConvertExpandoObjectToDictionary(dynParam);
        return jsRuntime.InvokeAsync<bool>($"{ChartJsInteropName}.setupChart", param);
    }

    /// <summary>
    /// This method is specifically used to convert an <see cref="ExpandoObject"/> with a Tree structure to a <c>Dictionary&lt;string, object&gt;</c>.
    /// </summary>
    /// <param name="expando">The <see cref="ExpandoObject"/> to convert.</param>
    /// <returns>The fully converted <see cref="ExpandoObject"/>.</returns>
    private static Dictionary<string, object> ConvertExpandoObjectToDictionary(ExpandoObject expando) => RecursivelyConvertIDictToDict(expando);

    /// <summary>
    /// This method takes an <c>IDictionary&lt;string, object&gt;</c> and recursively converts it to a <c>Dictionary&lt;string, object&gt;</c>.
    /// The idea is that every <c>IDictionary&lt;string, object&gt;</c> in the tree will be of type <c>Dictionary&lt;string, object&gt;</c> instead of some other implementation like <see cref="ExpandoObject"/>.
    /// </summary>
    /// <param name="value">The <c>IDictionary&lt;string, object&gt;</c> to convert</param>
    /// <returns>The fully converted <c>Dictionary&lt;string, object&gt;</c></returns>
    private static Dictionary<string, object> RecursivelyConvertIDictToDict(IDictionary<string, object> value) =>
        value.ToDictionary(
            keySelector => keySelector.Key,
            elementSelector =>
            {
                // if it's another IDict just go through it recursively
                if (elementSelector.Value is IDictionary<string, object> dict)
                {
                    return RecursivelyConvertIDictToDict(dict);
                }

                // if it's an IEnumerable check each element
                if (elementSelector.Value is IEnumerable<object> list)
                {
                    // go through all objects in the list
                    // if the object is an IDict -> convert it
                    // if not keep it as is
                    return list
                        .Select(o => o is IDictionary<string, object> dictionary
                            ? RecursivelyConvertIDictToDict(dictionary)
                            : o
                        );
                }

                // neither an IDict nor an IEnumerable -> it's fine to just return the value it has
                return elementSelector.Value;
            }
        );

    /// <summary>
    /// Update an existing chart. Make sure that the Chart with this <see cref="ConfigBase.CanvasId"/> already exists.
    /// </summary>
    /// <param name="jsRuntime"></param>
    /// <param name="chartConfig">The updated config of the chart you want to update.</param>
    /// <returns></returns>
    public static ValueTask<bool> UpdateChart(this IJSRuntime jsRuntime, ConfigBase chartConfig)
    {
        dynamic dynParam = StripNulls(chartConfig);
        Dictionary<string, object> param = ConvertExpandoObjectToDictionary(dynParam);
        return jsRuntime.InvokeAsync<bool>($"{ChartJsInteropName}.updateChart", param);
    }

    /// <summary>
    /// Returns an object that is equivalent to the given parameter but without any null members AND it preserves <see cref="IMethodHandler"/>s intact.
    /// <para>Preserving <see cref="IMethodHandler"/> members is important because they might be <see cref="DelegateHandler{T}"/> instances which contain
    /// delegates that can't be (de)serialized.</para>
    /// <para>Stripping null members is only needed because chartJs doesn't handle null values and undefined values the same and with JSRuntime null gets
    /// serialized to null instead of undefined (not at all) and WE CAN'T CHANGE THAT (see https://github.com/aspnet/AspNetCore/issues/12685).
    /// If this were not the case, no null member stripping were necessary -> no json.net serialize-deserialize magic -> no loss of <see cref="DelegateHandler{T}"/>
    /// instances -> no recovery of those. Everything would be better with AspNetCore#12685 finally being implemented but to fully migrate to System.Text.Json
    /// we might also need corefx#38650 and corefx#39905.
    /// Nevertheless, The Show must go on!</para>
    /// </summary>
    /// <param name="chartConfig">The config you want to strip of null members.</param>
    /// <returns></returns>
    private static ExpandoObject StripNulls(ConfigBase chartConfig)
    {
        // Serializing with the custom serializer settings remove null members
        string cleanChartConfigStr = JsonConvert.SerializeObject(chartConfig, JsonSerializerSettings);

        // Get back an ExpandoObject dynamic with the clean config - having an ExpandoObject allows us to add/replace members regardless of type
        ExpandoObject cleanChartConfig = JsonConvert.DeserializeObject<ExpandoObject>(cleanChartConfigStr, new ExpandoObjectConverter());

#pragma warning disable CS0618 // Type or member is obsolete

        // Restore any .net refs that need to be passed intact
        // TODO Find a way to do this dynamically. Maybe with attributes or something like that?
        dynamic dynamicChartConfig = (dynamic)chartConfig;
        if (dynamicChartConfig?.Options?.OnClick is IMethodHandler chartOnClick)
        {
            cleanChartConfig.SetValue(path: "options.onClick", chartOnClick);
        }

        if (dynamicChartConfig?.Options?.OnHover is IMethodHandler chartOnHover)
        {
            cleanChartConfig.SetValue(path: "options.onHover", chartOnHover);
        }

        if (dynamicChartConfig?.Options?.Legend?.OnClick is IMethodHandler legendOnClick)
        {
            cleanChartConfig.SetValue(path: "options.legend.onClick", legendOnClick);
        }

        if (dynamicChartConfig?.Options?.Legend?.OnHover is IMethodHandler legendOnHover)
        {
            cleanChartConfig.SetValue(path: "options.legend.onHover", legendOnHover);
        }

        if (dynamicChartConfig?.Options?.Legend?.Labels?.GenerateLabels is IMethodHandler generateLabels)
        {
            cleanChartConfig.SetValue(path: "options.legend.labels.generateLabels", generateLabels);
        }

        if (dynamicChartConfig?.Options?.Legend?.Labels?.Filter is IMethodHandler filter)
        {
            cleanChartConfig.SetValue(path: "options.legend.labels.filter", filter);
        }

        // Ticks callback need special handling because it can be either a single scale or two arrays of scales (xAxes and yAxes)
        // it's really ugly (and quite slow), I hope we can improve this later on. Also ms, PLEASE, give us customizable jsruntime serialization.
        try
        {
            if (dynamicChartConfig?.Options?.Scale?.Callback is IMethodHandler singleScaleTickCallback)
            {
                cleanChartConfig.SetValue(path: "options.scale.callback", singleScaleTickCallback);
            }
        }
        catch (Microsoft.CSharp.RuntimeBinder.RuntimeBinderException) // happens when the options don't have a Scale property
        {
            try
            {
                // here we trust that if the Scales property exists, it contains an XAxes and a YAxes property
                if (dynamicChartConfig?.Options?.Scales?.XAxes is IEnumerable<object> xAxes)
                {
                    AssignAxes(xAxes, "options.scales.xAxes");
                }

                if (dynamicChartConfig?.Options?.Scales?.YAxes is IEnumerable<object> yAxes)
                {
                    AssignAxes(yAxes, "options.scales.yAxes");
                }
            }
            catch (Microsoft.CSharp.RuntimeBinder.RuntimeBinderException) { } // happens when the options don't have a Scales property (only Pie & Doughnut)
        }

        return cleanChartConfig;

        void AssignAxes(IEnumerable<object> axes, string axesPath)
        {
            IEnumerable<object> axesInDynamic = cleanChartConfig.GetValue(axesPath) as IEnumerable<object>;

            foreach ((object axis, ExpandoObject axisInDynamic) in axes.Zip(axesInDynamic, (axis, axisInDynamic) => (axis, (ExpandoObject)axisInDynamic)))
            {
                if (((dynamic)axis)?.Ticks?.Callback is IMethodHandler axisTickCallback)
                {
                    axisInDynamic.SetValue("ticks.callback", axisTickCallback);
                }
            }
        }

#pragma warning restore CS0618 // Type or member is obsolete
    }
}
